## Exercise 12
Betul Ozturan

 Download the `indeed_scraped_data.zip` file from Canvas under `Files/datasets`. Unzip it and use all the files under the folder "job_info_data" that are scraped on May 22, 2022 (having "5222022" in the filename) for the following assignments. 

In [1]:
import pandas as pd
import re # regular expressions
import json

In [2]:
import spacy

In [5]:
file_list = os.listdir('indeed_scraped_data/job_info_data')
file_list.remove('.DS_Store') # remove DS_Store

files =[]
for i in range(len(file_list)):
    x= re.search(".*5222022",file_list[i])
    if x != None:
        files.append(file_list[i])

for i in range(len(files)): 
    files[i] = 'indeed_scraped_data/job_info_data/'+files[i]

## Task 1. 

 Create one pandas dataframe that combines all the data scraped from May 22, 2022 together. Drop rows with missing job titles and/or job descriptions. Use `spacy` to tokenize all the job titles included in the cleaned dataframe. For each job title, find all the nouns and all the adjectives in the title and get their lowercased lemmatized form. Use the reformatted nouns to construct a vocabulary set for this dataframe. How many unique nouns are there? Construct another vocabulary set using the reformatted adjectives. How many unique adjectives are there? What kind of different information do the nouns versus the adjectives reveal about the specific job? 


In [6]:
data_frame = pd.DataFrame()
content = []
  
# checking all the csv files in the 
# specified path
for filename in files:
    
    # reading content of csv file
    # content.append(filename)
    df = pd.read_csv(filename, index_col=None)
    content.append(df)
  
# converting content to data frame
data_frame = pd.concat(content)
data_frame = data_frame.dropna(subset=['lnks_job_title','lnks_job_description'])

In [8]:
data_frame= pd.DataFrame(data_frame).reset_index()
data_frame.head()

level_0  index                                          lnks_link  \
0        0    559  https://www.indeed.com/rc/clk?jk=069416ccc58dc...   
1        1    560  https://www.indeed.com/pagead/clk?mo=r&ad=-6NY...   
2        2    562  https://www.indeed.com/rc/clk?jk=263b35a70c9a8...   
3        3    563  https://www.indeed.com/rc/clk?jk=e15dbd7ec34e6...   
4        4    564  https://www.indeed.com/rc/clk?jk=d0e4f6dc1721c...   

                             lnks_job_title  \
0   Maintenance Controller (A&P) Technician   
1                           Program Manager   
2  Senior Hydrogeologist, Boulder, Colorado   
3             Contract Administration Mgr I   
4                     Sr. Quality Assurance   

                           lnks_company  \
0          Total Aviation Staffing, LLC   
1  Norton Staffing and Recruiting Group   
2               Formation Environmental   
3        Dept of Family & Protectve Svc   
4                                 Amgen   

                                    lnks_company_url lnks_company_location  \
0  https://www.indeed.com/cmp/Total-Aviation-Staf...  Washington, DC 20500   
1  https://www.indeed.com/cmp/Norton-Staffing-and...           Meriden, CT   
2  https://www.indeed.com/cmp/Formation-Environme...     Boulder, CO 80301   
3  https://www.indeed.com/cmp/Texas-Health-and-Hu...      Austin, TX 78751   
4  https://www.indeed.com/cmp/Amgen?campaignid=mo...   Cambridge, MA 02139   

                                lnks_job_description  
0  Do you want to be part of a world-class team w...  
1  BH: 3.1\n***Our Client is looking for a Progra...  
2  Senior Hydrogeologist\nFormation Environmental...  
3  Contract Administration Mgr I - THIS POISTION ...  
4  HOW MIGHT YOU DEFY IMAGINATION?\nYou’ve earned...

In [10]:
len(data_frame)

4803

In [9]:
nlp = spacy.load('en_core_web_sm')

In [83]:
job_title = data_frame['lnks_job_title']

For each job title, find all the nouns and all the adjectives in the title and get their lowercased lemmatized form. Use the reformatted nouns to construct a vocabulary set for this dataframe. How many unique nouns are there? Construct another vocabulary set using the reformatted adjectives. How many unique adjectives are there? What kind of different information do the nouns versus the adjectives reveal about the specific job? 

In [84]:
job_title_lower = []
for title in job_title:
    job_title_lower.append(title.lower())


In [21]:
token2tag = {}

for i in range(len(job_title)):
    job_title_token = job_title[i]
    for token in nlp(job_title_token):
    # print(token.text, token.pos_, token.dep_)
        token2tag[token.text] = token.pos_

token_df = pd.DataFrame(columns=['token', 'tag'])
token_df['token'] = token2tag.keys()
token_df['tag'] = token2tag.values()

print('number of unique tokens in the job description:', len(token2tag))
print('the distribution of token tags among them:')
print(token_df['tag'].value_counts())

number of unique tokens in the job description: 2539
the distribution of token tags among them:
PROPN    1980
NOUN      276
NUM        94
VERB       55
ADJ        47
ADP        22
PUNCT      18
ADV         8
CCONJ       6
PART        6
X           6
AUX         6
DET         4
SPACE       4
SYM         3
PRON        2
INTJ        2
Name: tag, dtype: int64


In [23]:
for title in job_title:
    doc = nlp(title)
    for chunk in doc.noun_chunks:
        print([
            chunk,            # A Span object with the full phrase.
            chunk.root,       # The key Token within this phrase.
            chunk.root.dep_,  # The grammatical role of this phrase.
            chunk.root.head   # The grammatical parent Token.
        ])

[Maintenance Controller, Controller, 'ROOT', Controller]
[A&P, A&P, 'appos', Controller]
[) Technician, Technician, 'appos', Controller]
[Program Manager, Manager, 'ROOT', Manager]
[Senior Hydrogeologist, Hydrogeologist, 'ROOT', Hydrogeologist]
[Boulder, Boulder, 'appos', Hydrogeologist]
[Colorado, Colorado, 'appos', Boulder]
[I, I, 'ROOT', I]
[Sr. Quality Assurance, Assurance, 'ROOT', Assurance]
[Lead Licensed Clinical Social Worker, Worker, 'ROOT', Worker]
[LCSW, LCSW, 'appos', Worker]
[Process Control Systems Specialist, Specialist, 'ROOT', Specialist]
[Business Attorney, Attorney, 'ROOT', Attorney]
[Licensed Practical Nurse, Nurse, 'ROOT', Nurse]
[(LPN)- Night Shift, Shift, 'appos', Nurse]
[Quality Assurance-Entry, Entry, 'ROOT', Entry]
[Business Relationships Manager, Manager, 'ROOT', Manager]
[HR Generalist, Generalist, 'ROOT', Generalist]
[Vice President, President, 'ROOT', President]
[Corporate Audit, Audit, 'appos', President]
[Security Operations Specialist, Specialist, 'ROOT

In [64]:
noun_title = []
for title in job_title_lower:
    doc = nlp(title)
    for chunk in doc.noun_chunks:
        if chunk.root.head.pos_ == 'NOUN':
        # print([chunk, chunk.root, chunk.root.dep_, chunk.root.head.lemma_])
            noun = (chunk.root.head.lemma_)
            noun_title.append(noun)
            print('noun:', noun)

noun: manager
noun: hydrogeologist
noun: hydrogeologist
noun: assurance
noun: worker
noun: worker
noun: specialist
noun: attorney
noun: nurse
noun: nurse
noun: entry
noun: manager
noun: generalist
noun: president
noun: president
noun: specialist
noun: member
noun: member
noun: stocker
noun: inventory
noun: service
noun: member
noun: director
noun: purchasing
noun: engineer
noun: technologist
noun: manager
noun: director
noun: manager
noun: chef
noun: analyst
noun: manager
noun: manager
noun: officer
noun: manager
noun: program
noun: supervisor
noun: consultant
noun: pilot
noun: supervisor
noun: member
noun: member
noun: member
noun: administrator
noun: hire
noun: lab
noun: secretary
noun: pharmacist
noun: engineer
noun: director
noun: judge
noun: member
noun: member
noun: member
noun: controller
noun: specialist
noun: director
noun: director
noun: accountant
noun: bk3
noun: construction
noun: assurance
noun: analyst
noun: administrator
noun: phlebotomy
noun: representative
noun: repres

In [45]:
len(noun_title)

23

In [41]:
adj_title = []
for title in job_title_lower:
    doc = nlp(title)
    for chunk in doc.noun_chunks:
        if chunk.root.head.pos_ == 'ADJ':
        # print([chunk, chunk.root, chunk.root.dep_, chunk.root.head.lemma_])
            adj = (chunk.text, chunk.root.head.lemma_)
            adj_title.append(adj)
            print('adj:', adj)

adj: ('housekeeping', 'clean')
adj: ('business analyst', 'regulatory')
adj: ('it', 'regulatory')
adj: ('az', 'remote')
adj: ('import/export specialist', 'onsite')
adj: ('import/export specialist', 'onsite')
adj: ('sanctions', 'aml')
adj: ('housekeeping', 'clean')
adj: ('time', 'butcher)-full')
adj: ('manager', 'regulatory')
adj: ('service delivery associate', 'remote')
adj: ('lead', 'dietitian')
adj: ('physical therapist - part', 'standish')
adj: ('import/export specialist', 'onsite')
adj: ('import/export specialist', 'onsite')
adj: ('housekeeping', 'clean')
adj: ('quality operations', 'professional')
adj: ('preferred lending credit solutions', 'advisor-')
adj: ('north kingstown financial center- north kingstown', 'advisor-')
adj: ('teller', 'main')
adj: ('senior business services', 'external')
adj: ('service delivery associate', 'remote')
adj: ('outpatient', 'inpatient')


In [ ]:
len(adj_title)

The adjectives give clues about the nature of the job such as remote vs in person. While nouns give us the concept of the jobsuch as specialist, manager etc. 

## Task 2. 

Task 2. Choose the first job title in your dataframe as the primary string. Use one-hot encoding as the word embedding method and find jobs in your cleaned dataframe that have similar nouns in the title as your primary string. 

In [114]:
import numpy as np

title_a = job_title_lower[0] # take one example first
token_indices = []

for token in nlp(title_a):
    if token.is_alpha:
        print(token.lemma_.lower())
        if token.lemma_.lower() in noun_title:
            token_index_in_vocab = noun_title.index(token.lemma_.lower())
            token_indices.append(token_index_in_vocab)
        else:
            print('token not recognized')

one_hot_encoding = np.zeros(len(noun_title))
for token_index in token_indices:
    one_hot_encoding[token_index] = 1
print(one_hot_encoding)

maintenance
controller
technician
[0. 0. 0. ... 0. 0. 0.]


In [115]:
one_hot_encodings = []

for i in range(len(job_title_lower)):
    job_title = job_title_lower[0] # take one example first
    token_indices = []

    one_hot_encoding = np.zeros(len(noun_title))
    for token_index in token_indices:
        one_hot_encoding[token_index] = 1

    one_hot_encodings.append(one_hot_encoding)

In [109]:
from scipy.spatial.distance import cosine

cosine(one_hot_encodings[11], one_hot_encodings[10]) # cosine distance
1 - cosine(one_hot_encodings[11], one_hot_encodings[10]) # cosine_similarity

1

In [112]:
title_a = job_title_lower[0]
similarity_values = []
for i in range(1, len(job_title_lower)):
    similarity_value = 1 - cosine(one_hot_encodings[0], one_hot_encodings[i])
    similarity_values.append(similarity_value)


/Users/betulozturan/Library/Python/3.9/lib/python/site-packages/scipy/spatial/distance.py:620: RuntimeWarning: invalid value encountered in double_scalars
  dist = 1.0 - uv / np.sqrt(uu * vv)


In [113]:
similarity_values

[1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,
 1,


In [72]:
similar_df = pd.DataFrame(columns=['job_title', 'similarity_value_with_one_hot'])
similar_df['job_title'] = job_df.loc[1:, 'lnks_job_title']
similar_df['similarity_value_with_one_hot'] = similarity_values
print('the primary job title:', title_a)
similar_df.sort_values(by='similarity_value_with_one_hot', ascending=False)


the primary job title: Maintenance Controller (A&P) Technician


job_title  \
1                                       Program Manager   
3208  Travel - Interim Director of Nursing - Long Te...   
3206  Accountant I (Accountant II consideration)-Red...   
3205  Registered Nurse - Supervisor - Union West Sur...   
3204                                   Systems Engineer   
...                                                 ...   
1601  Travel - Interim Director of Nursing - Long Te...   
1600                        Pharmacy Operations Manager   
1599    Financial Crimes - Sanctions - Senior Associate   
1598                        2022-23 Vocal Music Teacher   
4802                                     Grants Manager   

      similarity_value_with_one_hot  
1                                 1  
3208                              1  
3206                              1  
3205                              1  
3204                              1  
...                             ...  
1601                              1  
1600                              1  
1599                              1  
1598                              1  
4802                              1  

[4802 rows x 2 columns]

When I only use noun vocabulary, all of the titles are calculated as same. This is how I understood the question but the results do not seem good. 

### Task 3. 
Use spacy's word vector to do Task 2. Compare the results. 

In [166]:
# Similarity of two documents

similarity = []
for title in job_title_lower:
    doc1 = nlp(job_title_lower[0]) 
    doc2 = nlp(title)
    similarity.append(
        {"doc1" : doc1, 
        "doc2": doc2, 
        "similarity": doc1.similarity(doc2)
        }
    )
    print(doc1, "<->", doc2, doc1.similarity(doc2))


maintenance controller (a&p) technician <-> maintenance controller (a&p) technician 1.0
maintenance controller (a&p) technician <-> program manager 0.44339574500113854
maintenance controller (a&p) technician <-> senior hydrogeologist, boulder, colorado 0.44004644778004803
maintenance controller (a&p) technician <-> contract administration mgr i 0.28355940265499735
maintenance controller (a&p) technician <-> sr. quality assurance 0.4520766044369361
maintenance controller (a&p) technician <-> lead licensed clinical social worker (lcsw) 0.48726035040598403
maintenance controller (a&p) technician <-> process control systems specialist 0.5818815880501031
maintenance controller (a&p) technician <-> business attorney 0.37721664400893706
maintenance controller (a&p) technician <-> licensed practical nurse (lpn)- night shift 0.47497226079759336
maintenance controller (a&p) technician <-> quality assurance-entry 0.2775758518536071
maintenance controller (a&p) technician <-> business relationship

/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/2333053001.py:10: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  "similarity": doc1.similarity(doc2)
/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/2333053001.py:13: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use c

maintenance controller (a&p) technician <-> branch operations manager 0.4612145206555023
maintenance controller (a&p) technician <-> assistant chef at bradfield place 0.43407650380306323
maintenance controller (a&p) technician <-> personal consultation associate analyst 0.4450778204838771
maintenance controller (a&p) technician <-> ehs manager 0.4185260484488072
maintenance controller (a&p) technician <-> dairy plant manager 0.4022949831883895
maintenance controller (a&p) technician <-> mortgage loan officer - peoria, az 0.6341476663793125
maintenance controller (a&p) technician <-> fire life safety officer 0.37948396732509454
maintenance controller (a&p) technician <-> information systems compliance manager 0.4591596218636326
maintenance controller (a&p) technician <-> compliance program advisor- international compliance program 0.4703617519736612
maintenance controller (a&p) technician <-> abatement supervisor 0.37890267849516807
maintenance controller (a&p) technician <-> senior con

In [168]:
similarity = pd.DataFrame(similarity)
similarity.sort_values(by='similarity', ascending=False)


doc1  \
0     (maintenance, controller, (, a&p, ), technician)   
4397  (maintenance, controller, (, a&p, ), technician)   
1944  (maintenance, controller, (, a&p, ), technician)   
1805  (maintenance, controller, (, a&p, ), technician)   
2071  (maintenance, controller, (, a&p, ), technician)   
...                                                ...   
4396  (maintenance, controller, (, a&p, ), technician)   
1631  (maintenance, controller, (, a&p, ), technician)   
2181  (maintenance, controller, (, a&p, ), technician)   
4363  (maintenance, controller, (, a&p, ), technician)   
1643  (maintenance, controller, (, a&p, ), technician)   

                                                  doc2  similarity  
0     (maintenance, controller, (, a&p, ), technician)    1.000000  
4397  (maintenance, controller, (, a&p, ), technician)    1.000000  
1944  (maintenance, controller, (, a&p, ), technician)    1.000000  
1805  (maintenance, controller, (, a&p, ), technician)    1.000000  
2071  (maintenance, controller, (, a&p, ), technician)    1.000000  
...                                                ...         ...  
4396                                             (lpn)   -0.005573  
1631                                             (lpn)   -0.005573  
2181                                      (dispatcher)   -0.008099  
4363                   (meat, and, produce, associate)   -0.082010  
1643                   (meat, and, produce, associate)   -0.082010  

[4803 rows x 3 columns]

Spacy gives a much more accurate description and shows us a spectrum of similarity. 

Bonus task 1. Do task 3 for both the nouns and the adjectives in the title. Combine the similarity values from comparing the nouns and from comparing the adjectives, and find jobs in your cleaned dataframe that are similar to your primary string in the combined setting. Compare the results with what you get from Task 3.  [0.5 extra point] 

In [137]:
similarity_noun = []
for title in noun_title:
    doc1 = nlp(noun_title[0])
    doc2 = nlp(title)
    similarity_noun.append(
        {"doc1" : doc1, 
        "doc2": doc2, 
        "similarity": doc1.similarity(doc2)
        }
    )
    print(doc1, "<->", doc2, doc1.similarity(doc2))


manager <-> manager 1.0
manager <-> hydrogeologist 0.37206089798607117
manager <-> hydrogeologist 0.37206089798607117
manager <-> assurance 0.390704834812081
manager <-> worker 0.7731646644902618
manager <-> worker 0.7731646644902618
manager <-> specialist 0.661934570117932
manager <-> attorney 0.6643961185026093
manager <-> nurse 0.6672223922624856
manager <-> nurse 0.6672223922624856
manager <-> entry 0.5803142169402528
manager <-> manager 1.0
manager <-> generalist 0.654699214037459
manager <-> president 0.7823504940865841
manager <-> president 0.7823504940865841
manager <-> specialist 0.661934570117932
manager <-> member 0.789761800664654
manager <-> member 0.789761800664654
manager <-> stocker 0.7570489856407888
manager <-> inventory 0.4723562313671913
manager <-> service 0.6720094039622913
manager <-> member 0.789761800664654
manager <-> director 0.8533332024446092
manager <-> purchasing 0.2200716768614563
manager <-> engineer 0.5808952816063881
manager <-> technologist 0.7126468

/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/1616349184.py:8: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  "similarity": doc1.similarity(doc2)
/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/1616349184.py:11: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use co

manager <-> manager 1.0
manager <-> director 0.8533332024446092
manager <-> manager 1.0
manager <-> chef 0.7356979232799209
manager <-> analyst 0.8168070772133468
manager <-> manager 1.0
manager <-> manager 1.0
manager <-> officer 0.7588560466900914
manager <-> manager 1.0
manager <-> program 0.6803238612992624
manager <-> supervisor 0.6645742969699878
manager <-> consultant 0.6995559979185043
manager <-> pilot 0.7261381083317616
manager <-> supervisor 0.6645742969699878
manager <-> member 0.789761800664654
manager <-> member 0.789761800664654
manager <-> member 0.789761800664654
manager <-> administrator 0.7495474437115377
manager <-> hire 0.21192488877841165
manager <-> lab 0.5048789346094149
manager <-> secretary 0.6934532158891632
manager <-> pharmacist 0.664455956550351
manager <-> engineer 0.5808952816063881
manager <-> director 0.8533332024446092
manager <-> judge 0.6977678543496069
manager <-> member 0.789761800664654
manager <-> member 0.789761800664654
manager <-> member 0.78

When I use only nouns, the degree of similarity changes: for instance 
program manager <-> senior hydrogeologist, boulder, colorado 0.13090587175958365 vs manager <-> hydrogeologist 0.37206089798607117

In [143]:
similarity_adj = []
for title in adj_title:
    doc1 = nlp(adj_title[0][1])
    doc2 = nlp(title[1])
    similarity_adj.append(
        {"doc1" : doc1, 
        "doc2": doc2, 
        "similarity": doc1.similarity(doc2)
        }
    )
    print(doc1, "<->", doc2, doc1.similarity(doc2))

clean <-> clean 1.0
clean <-> regulatory 0.6802749831153856
clean <-> regulatory 0.6802749831153856
clean <-> remote 0.7083923664155802
clean <-> onsite 0.5091618757654325
clean <-> onsite 0.5091618757654325
clean <-> aml 0.49267109391800107
clean <-> clean 1.0
clean <-> butcher)-full 0.6708305271548883
clean <-> regulatory 0.6802749831153856
clean <-> remote 0.7083923664155802
clean <-> dietitian 0.5193148015082588
clean <-> standish 0.6552330383115208
clean <-> onsite 0.5091618757654325
clean <-> onsite 0.5091618757654325
clean <-> clean 1.0
clean <-> professional 0.7326943931993405
clean <-> advisor- 0.2650578158345522
clean <-> advisor- 0.2650578158345522
clean <-> main 0.6226610944465234
clean <-> external 0.6256834135589925
clean <-> remote 0.7083923664155802
clean <-> inpatient 0.7765711765592577


/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/3105604143.py:8: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  "similarity": doc1.similarity(doc2)
/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/3105604143.py:11: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use co

I am working with a much limited data in adjectives. It seems that task 3 results are much more accurate than using only nouns or adjectives. 

Bonus task 2. Do task 3 for job descriptions. Compare the results with what you get from processing job titles. Which data source gives you better results in terms of finding actual similar jobs? Can you guess why? [0.5 extra point] 

You can also try doing Bonus task 2 using (1) verbs only; (2) noun-verb pairs, and compare their results with using nouns only. How are the results different from each other? 

In [148]:
similarity_desc = []
job_description = data_frame["lnks_job_description"]

for description in job_description:
    doc1 = nlp(job_description[0].lower()) 
    doc2 = nlp(description.lower())
    similarity_desc.append(
        {"doc1" : doc1, 
        "doc2": doc2, 
        "similarity": doc1.similarity(doc2)
        }
    )


/var/folders/zf/mq4rgmxx1gg0xk4czpw0qswh0000gn/T/ipykernel_16566/704476736.py:10: UserWarning: [W007] The model you're using has no word vectors loaded, so the result of the Doc.similarity method will be based on the tagger, parser and NER, which may not give useful similarity judgements. This may happen if you're using one of the small models, e.g. `en_core_web_sm`, which don't ship with word vectors and only use context-sensitive tensors. You can always add your own word vectors, or use one of the larger models instead if available.
  "similarity": doc1.similarity(doc2)


In [150]:
similarity_desc = pd.DataFrame(similarity_desc)

In [153]:
similarity_desc 

doc1  \
0     (do, you, want, to, be, part, of, a, world, -,...   
1     (do, you, want, to, be, part, of, a, world, -,...   
2     (do, you, want, to, be, part, of, a, world, -,...   
3     (do, you, want, to, be, part, of, a, world, -,...   
4     (do, you, want, to, be, part, of, a, world, -,...   
...                                                 ...   
4798  (do, you, want, to, be, part, of, a, world, -,...   
4799  (do, you, want, to, be, part, of, a, world, -,...   
4800  (do, you, want, to, be, part, of, a, world, -,...   
4801  (do, you, want, to, be, part, of, a, world, -,...   
4802  (do, you, want, to, be, part, of, a, world, -,...   

                                                   doc2  similarity  
0     (do, you, want, to, be, part, of, a, world, -,...    1.000000  
1     (bh, :, 3.1, \n, *, *, *, our, client, is, loo...    0.885568  
2     (senior, hydrogeologist, \n, formation, enviro...    0.918361  
3     (contract, administration, mgr, i, -, this, po...    0.965163  
4     (how, might, you, defy, imagination, ?, \n, yo...    0.969887  
...                                                 ...         ...  
4798  (the, mission, is, to, encourage, economic, gr...    0.952864  
4799  (associate, director, of, advancement, service...    0.938926  
4800  (objective, :, responsible, for, processing, a...    0.967588  
4801  (your, role, as, a, store, scan, specialist, w...    0.945065  
4802  (grants, manager, \n, reports, to, :, executiv...    0.961304  

[4803 rows x 3 columns]

In [169]:
sim = pd.concat([similarity_desc, similarity], axis = 1)

In [170]:
sim

doc1  \
0     (do, you, want, to, be, part, of, a, world, -,...   
1     (do, you, want, to, be, part, of, a, world, -,...   
2     (do, you, want, to, be, part, of, a, world, -,...   
3     (do, you, want, to, be, part, of, a, world, -,...   
4     (do, you, want, to, be, part, of, a, world, -,...   
...                                                 ...   
4798  (do, you, want, to, be, part, of, a, world, -,...   
4799  (do, you, want, to, be, part, of, a, world, -,...   
4800  (do, you, want, to, be, part, of, a, world, -,...   
4801  (do, you, want, to, be, part, of, a, world, -,...   
4802  (do, you, want, to, be, part, of, a, world, -,...   

                                                   doc2  similarity  \
0     (do, you, want, to, be, part, of, a, world, -,...    1.000000   
1     (bh, :, 3.1, \n, *, *, *, our, client, is, loo...    0.885568   
2     (senior, hydrogeologist, \n, formation, enviro...    0.918361   
3     (contract, administration, mgr, i, -, this, po...    0.965163   
4     (how, might, you, defy, imagination, ?, \n, yo...    0.969887   
...                                                 ...         ...   
4798  (the, mission, is, to, encourage, economic, gr...    0.952864   
4799  (associate, director, of, advancement, service...    0.938926   
4800  (objective, :, responsible, for, processing, a...    0.967588   
4801  (your, role, as, a, store, scan, specialist, w...    0.945065   
4802  (grants, manager, \n, reports, to, :, executiv...    0.961304   

                                                  doc1  \
0     (maintenance, controller, (, a&p, ), technician)   
1     (maintenance, controller, (, a&p, ), technician)   
2     (maintenance, controller, (, a&p, ), technician)   
3     (maintenance, controller, (, a&p, ), technician)   
4     (maintenance, controller, (, a&p, ), technician)   
...                                                ...   
4798  (maintenance, controller, (, a&p, ), technician)   
4799  (maintenance, controller, (, a&p, ), technician)   
4800  (maintenance, controller, (, a&p, ), technician)   
4801  (maintenance, controller, (, a&p, ), technician)   
4802  (maintenance, controller, (, a&p, ), technician)   

                                                   doc2  similarity  
0      (maintenance, controller, (, a&p, ), technician)    1.000000  
1                                    (program, manager)    0.443396  
2     (senior, hydrogeologist, ,, boulder, ,, colorado)    0.440046  
3                    (contract, administration, mgr, i)    0.283559  
4                           (sr, ., quality, assurance)    0.452077  
...                                                 ...         ...  
4798                                       (controller)    0.376803  
4799   (associate, director, of, advancement, services)    0.496202  
4800  (treasury, management, ach, &, wire, operation...    0.550567  
4801  (store, scanning, specialist, (, pricing, inte...    0.582491  
4802                                  (grants, manager)    0.496556  

[4803 rows x 6 columns]

When we compare the descriptions the degree of similarity increases a lot. 